cvat/apps/dataset_manager/bindings.py

```diff
@@ -2151,10 +2151,62 @@ def reduce_fn(acc, v):
                 raise CvatImportError("Image {}: can't import annotation "
                     "#{} ({}): {}".format(item.id, idx, ann.type.name, e)) from e
 
-    for track in tracks.values():
-        track['elements'] = list(track['elements'].values())
-        instance_data.add_track(instance_data.Track(**track))
+    def _validate_track_shapes(shapes):
+        new_shapes = []
+        prev_shape = None
+        # infer the keyframe shapes and keep only them
+        for shape in shapes:
+            cur_is_visible = shape is not None and not shape.outside
+            if not cur_is_visible:
+                prev_shape = shape
+                continue
+
+            prev_is_visible = prev_shape is not None and not prev_shape.outside
+            has_gap = False
+            if not prev_is_visible or (
+                has_gap := prev_shape.frame + instance_data.frame_step < shape.frame
+            ) or shape.keyframe:
+                if has_gap:
+                    prev_shape = prev_shape._replace(outside=True, keyframe=True,
+                        frame=prev_shape.frame + instance_data.frame_step)
+                    new_shapes.append(prev_shape)
 
+                shape = shape._replace(keyframe=True)
+                new_shapes.append(shape)
+
+            prev_shape = shape
```

Comment:
This algorithm misses some required outside shapes, please check the test task

```suggestion
        for shape in shapes:
            prev_is_visible = prev_shape and not prev_shape.outside
            cur_is_visible = shape and not shape.outside
            if not cur_is_visible:
                if prev_is_visible:
                    shape = shape._replace(keyframe=True)
                    new_shapes.append(shape)
                prev_shape = shape
                continue

            has_gap = False
            if prev_is_visible:
                has_gap = prev_shape.frame + instance_data.frame_step < shape.frame

            if not prev_is_visible or has_gap or shape.keyframe:
                if has_gap:
                    prev_shape = prev_shape._replace(outside=True, keyframe=True,
                        frame=prev_shape.frame + instance_data.frame_step)
                    new_shapes.append(prev_shape)

                shape = shape._replace(keyframe=True)
                new_shapes.append(shape)

            prev_shape = shape
```